signac/contrib/_index.py

```diff
@@ -0,0 +1,299 @@
+# Copyright (c) 2022 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""An optimized class for indexing signac Projects."""
+
+import json
+import logging
+from numbers import Number
+
+from ..errors import InvalidKeyError
+from .collection import (
+    _INDEX_OPERATORS,
+    _check_logical_operator_argument,
+    _DictPlaceholder,
+    _find_with_index_operator,
+    _float,
+    _TypedSetDefaultDict,
+    _valid_filter,
+)
+from .utility import _nested_dicts_to_dotted_keys, _to_hashable
+
+logger = logging.getLogger(__name__)
+
+_PRIMARY_KEY = "_id"
+
+
+class _Index(dict):
+    """A collection of documents, optimized for minimal use cases in signac.
+
+    The _Index class manages a collection of documents in memory.
+    A document is defined as a dictionary mapping of key-value pairs. Each
+    document is identified by a unique id. The _Index class is a
+    mapping from ids to documents, inheriting from :py:class:`dict`.
+
+    An instance of _Index may be used to manage and search documents.
+    For example, given a collection with member data, where each document
+    contains a `name` entry and an `age` entry, we can find the name of
+    all members that are at age 32 like this:
+
+    .. code-block:: python
+
+        members = {
+            '0': {'name': 'John',  'age': 32},
+            '1': {'name': 'Alice', 'age': 28},
+            '2': {'name': 'Kevin', 'age': 32},
+            # ...
+        }
+
+        member_collection = _Index(members)
+        for doc in member_collection.find({'age': 32}):
+            print(doc)  # prints 0 and 2
+
+    Parameters
+    ----------
+    docs : iterable
+        Initialize the collection with these documents.
+
+    """
+
+    def build_index(self, key):
+        """Build index for a given key.
+
+        This is a highly performance critical code path.
+
+        Parameters
+        ----------
+        key : str
+            The key on which the index is built.
+
+        Returns
+        -------
+        :class:`~_TypedSetDefaultDict`
+            Index for key.
+
+        Raises
+        ------
+        :class:`~signac.errors.InvalidKeyError`
+            The document contains invalid keys.
+
+        """
+        logger.debug(f"Building index for key '{key}'...")
+        nodes = key.split(".")
```

Comment:
AFAICT this function is almost exclusively called internally except for a single usage in schema code. Could we change the input to be a list of strings already split? That would remove an unnecessary join and resplit that occurs when this is called from `_find_expressions`.

Reply:
Hmm. I was able to eliminate one join by re-using the input key in the error check instead of computing `nodes.join(".")` and I made some other related improvements. See f4a05d415c3ffd55f18bfeaa9d3711b478a34dae. However, I'm not sure if we want to pass in `nodes` directly... I think that is more awkward and I don't see the performance benefit. We have to check if the original dotted key is present in the document and raise if so. Also I don't think it's worth changing the API. Maybe we can do that in a follow-up PR if you're still sure this would be helpful?

This microbenchmark in particular is kind of interesting: https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/nested_access.py I see that `recurse_all` is about 30% faster than `one_direct_access` when dealing with 2 levels of nesting, and about 40% faster when dealing with no nesting, so I changed our access pattern.

Reply:
My goal was mainly to avoid the redundant work. I don't feel strongly that we need to change the API as long as we can minimize the number of redundant string operations there.

That is interesting. It makes sense in hindsight, but I wouldn't have thought about it beforehand.